presto-druid/src/main/java/com/facebook/presto/druid/segment/SmooshedColumnSource.java

```diff
@@ -0,0 +1,130 @@
+    private static final String SMOOSH_METADATA_FILE_NAME = makeMetaFileName();
+    private static final String VERSION_FILE_NAME = "version.bin";
+    private static final String INDEX_METADATA_FILE_NAME = "index.drd";
+
+    private static final ObjectMapper JSON_MAPPER = new DefaultObjectMapper();
+    private static final SerializerUtils SERIALIZER_UTILS = new SerializerUtils();
+
+    private final IndexFileSource indexFileSource;
+    private final Map<String, SmooshFileMetadata> columnSmoosh = new TreeMap<>();
+
+    private int numFiles;
+
+    public SmooshedColumnSource(IndexFileSource indexFileSource)
+    {
+        this.indexFileSource = requireNonNull(indexFileSource, "indexFileSource is null");
+        loadSmooshFileMetadata();
+    }
+
+    @Override
+    public int getVersion()
+    {
+        try {
+            return ByteBuffer.wrap(indexFileSource.readFile(VERSION_FILE_NAME)).getInt();
+        }
+        catch (IOException e) {
+            throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, e);
+        }
+    }
+
+    @Override
+    public byte[] getColumnData(String name)
+    {
+        SmooshFileMetadata metadata = columnSmoosh.get(name);
+        if (metadata == null) {
+            throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Internal file %s doesn't exist", name));
+        }
+        String fileName = makeChunkFileName(metadata.getFileNum());
+        int fileStart = metadata.getStartOffset();
+        int fileSize = metadata.getEndOffset() - fileStart;
+
+        byte[] buffer = new byte[fileSize];
+        try {
+            indexFileSource.readFile(fileName, fileStart, buffer);
+        }
+        catch (IOException e) {
+            throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, e);
+        }
+        return buffer;
+    }
+
+    private void loadSmooshFileMetadata()
+    {
+        try {
+            byte[] metadata = indexFileSource.readFile(SMOOSH_METADATA_FILE_NAME);
+            BufferedReader in = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(metadata)));
+            String line = in.readLine();
+            if (line == null) {
+                throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Malformed metadata file: first line should be version,maxChunkSize,numChunks, got null."));
+            }
+
+            String[] splits = line.split(",");
+            if (!"v1".equals(splits[0])) {
+                throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Malformed metadata file: unknown version[%s], v1 is all I know.", splits[0]));
+            }
+            if (splits.length != 3) {
+                throw new PrestoException(DRUID_SEGMENT_LOAD_ERROR, format("Malformed metadata file: wrong number of splits[%d] in line[%s]", splits.length,
+                        line));
+            }
+            numFiles = Integer.valueOf(splits[2]);
+
+            while ((line = in.readLine()) != null) {
```

Comment:
I personally just prefer to have something like 
```java
while (true) {
      line = in.readLine();
      if (line == null) {
          break;
      }
......
}
```

But it's just personal preference and not strong opinion here :)